### DISCUSSION TITLE
Number and order of evaluations of eval_f and friends

### DISCUSSION THREAD

**Discussion by eike-fokken** [coin-or/Ipopt#532] [Q&A] [answered]
Hi there!

I plan to use Ipopt in the context of adjoint method optimization.
Therefore I have variables `x_0,...x_N` that map to dependent variables `y_0,...,y_K` and evaluating these is much more costly than evaluating the cost and constraints along with their derivatives.
Therefore I write a cache for the map `x |-> y`. 

My question is: Is it useful to have multiple entries in my cache? Or can I rely on the behaviour that after a certain `x` is used, all eval_f, eval_g and their jacobians are called once and only after that a new `x` is needed.

**Comment by svigerske** [comment, marked as answer]
I think it is unlikely that Ipopt will return to a previous iterate.

But it is not certain that Ipopt will request a Jacobian for every point where it called eval_f and/or eval_g. I could imagine that in the line search, for a trial point, first only objective and constraint function values are required. If the point is already rejected based on objective function value and constraint violation, Jacobian or Hessian shouldn't be required. See Section 2.3 in the implementation paper for details.

Further, in feasibility restoration phase, eval_f may not be called for a point at all.

**Reply by eike-fokken** [reply]
Thanks!
